### ISSUE TITLE
Tabular Neural Network hanging non-deterministically

### ISSUE THREAD

**Issue by Innixma** [awslabs/autogluon#149]
Defect introduced recently sometime around Open Source. I have disabled NN by default in Tabular until we find the cause and fix it.

**Comment by zhanghang1989** [comment]
Thanks for opening the issue tracker! You mentioned it was a recent change causing this issue, right? Do you know which version was working okay? I can compare the diff to debug that.

**Comment by Innixma** [comment]
Defect introduced on PR 94 due to adding pynvml dependency. For some reason pynvml causes tabular NN to permanently hang. Removing pynvml.

**Comment by Innixma** [comment]
Opened #153 to fix

**Comment by Innixma** [comment]
Merged

**Comment by austinmw** [comment]
Hi, I'm also experiencing this NN hanging, in SageMaker using the MXNet DLAMI's as base images. Doesn't look like they have pynvml installed. This happens with both CPU and GPU images:

`197237853439.dkr.ecr.us-east-2.amazonaws.com/autogluon-tabular-data-cpu-py36:latest` and
`197237853439.dkr.ecr.us-east-2.amazonaws.com/autogluon-tabular-data-gpu-py36-cu101:latest`

To test, I've been trying different `hyperparameter` options. Any time I use `NN` only, the predict calls hang. 

My local environment doesn't experience this issue, so I'm currently trying to tease out the difference(s). I'll update if I find anything. In case anyone gets a chance to look into the issue more, I'll attach the pip freeze of the cpu/gpu environments where it's failing.
[cpu_freeze.txt](https://github.com/awslabs/autogluon/files/4276885/cpu_freeze.txt)
[gpu_freeze.txt](https://github.com/awslabs/autogluon/files/4276887/gpu_freeze.txt)

**Comment by Innixma** [comment]
@austinmw This may be related to #345. Uncertain at this time though, as your issue is only on SageMaker. If it persists after #345 is merged, please make a new issue for this and we will look into it.

In your cpu_freeze.txt, you have LightGBM 2.3.0, but it should be 2.3.1 if you are using latest AutoGluon in mainline. Did you pip install (not fixed) or build from latest source (fixed)?

**Comment by austinmw** [comment]
Oops, good catch, had LightGBM 2.3.0 in my requirements file. Same issue with latest versions though. Will wait for that merge and check again, thanks.

**Comment by Innixma** [comment]
I'd recommend considering removing your requirements file altogether to ensure AutoGluon is installed with its preferred dependencies.

As for GPU, we have #263 . Currently Tabular does not support GPU but it could likely achieve significant speedups in areas you mentioned if the functionality was added. Feature pruning is an interesting use-case, but would require some more thought on how it could leverage GPUs.

**Comment by austinmw** [comment]
Tested with latest master and still hanging with NN. May be out of scope here since it's only happening in sagemaker, but will continue trying to figure out what's causing it.

**Comment by Innixma** [comment]
Assuming this is resolved at this point. If not, please re-open.
